### ISSUE TITLE
Query from DB only properties that declared in the projection

### ISSUE THREAD

**Issue by postoronnimw** [OData/WebApi#559]
I have domain model something like this:

```C#
public class TestEntity
{
    public virtual long Id { get; protected set; }
    public virtual TestDerivedSubEntity SubEntity { get; set; }
}
public class TestSubEntity
{
    public virtual long Id { get; protected set; }
    public virtual string Code { get; set; }
}
public class TestDerivedSubEntity : TestSubEntity
{
    public virtual string Name { get; set; }
	...
}
```

I execute following query from client:

```C#
var query = _dataContext.TestEntities.Select(c => new {c.Id, c.SubEntity.Code});
```

wich builds the following request:

```
http://localhost:51000/ResearchService/TestEntities?$expand=SubEntity($select=Code)&$select=Id
```

In built expression i see property

```C#
IsNull = $var1.SubEntity == null
```

Consequently in sql query (with NHibernate mapper) i see all properties of the TestDerivedSubEntity (not only Code):

```SQL
select testentity0_.id as col_0_0_,
	 case when testsubent1_.e_type='Derived' then 1 else 0 end as col_1_0_,
	 testentity0_.id_subentity as col_2_0_,
	 testsubent1_.code as col_3_0_,
	 testsubent1_.id as col_4_0_,
	 testsubent1_.id as id1_,
	 testsubent1_.code as code1_,
	 testsubent1_.name as name1_,
	 testsubent1_.e_type as e2_1_ 
from t_entity testentity0_ 
left outer join t_subentity testsubent1_ 
	on testentity0_.id_subentity=testsubent1_.id
```

I have deleted check `IsNull = $var1.SubEntity == null` from expression and now sql query contains only needed properties.
Also i noticed that comment on `NamedPropertyExpression.NullCheck` contains text about EntityFramework specifics:

```C#
// Checks whether this property is null or not. This is required for expanded navigation properties that are null as entityframework cannot
// create null's of type SelectExpandWrapper<ExpandedProperty> i.e. an expression like 
//       => new NamedProperty<Customer> { Value = order.Customer == null : null : new SelectExpandWrapper<Customer> { .... } } 
// cannot be translated by EF. So, we generate the following expression instead,
//       => new ExpandProperty<Customer> { Value = new SelectExpandWrapper<Customer> { .... }, IsNull = nullCheck }
// and use Value only if IsNull is false.
public Expression NullCheck { get; set; }
```

Which is there a correct way to select only need properties? Is it specifically for NHibernate only?

**Comment by congysu** [comment]
Currently, the SelectExpandBinder has an implicit dependency on EF.  We would like to introduce another layer of OData Abstract Syntax Tree (AST) binding to LINQ in vNext.  Your contribution to fix the specific issue is also very welcome.  OData Web API has code logic to detect specific LINQ provider, such as in `HandleNullPropagationOptionHelper`:
https://github.com/OData/WebApi/blob/f4573605bd88dabe882c4dc45e4a4ed0e55b32e0/OData/src/System.Web.OData/OData/Query/HandleNullPropagationOptionHelper.cs#L10

**Comment by VikingsFan** [comment]
Merged 14db4e498b46f810cde37f64bdf3451608fe575b
